### PR TITLE
New Task: Argo CD

### PR DESCRIPTION
**Pull request by AdityaGupta1**
# Changes

Created a `Task` that syncs (deploys) an Argo CD application. 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by AdityaGupta1** [comment]
I signed the CLA

File: argocd/argocd.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        description: name of the application to sync
+      - name: revision
+        description: the revision to sync to
+        default: HEAD
+      - name: flags
+        default: --
+  steps:
+    - name: login
+      image: argoproj/argocd
+      command: ["bash", "-c", "if [ ${inputs.params.token} = 'none' ]; \
```
Comment (vdemeester) [inline argocd/argocd.yaml]:
```
command: ["bash", "-c"]
args: - |
if [ …]
…
```

File: argocd/argocd.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        then yes | argocd login ${inputs.params.server} --username=${inputs.params.username} --password=${inputs.params.password}; \
+        else echo 'export ARGOCD_SERVER=${inputs.params.server} ; export ARGOCD_AUTH_TOKEN=${inputs.params.token}' > ~/env.sh; \
+        chmod 777 ~/env.sh; fi"]
+    - name: sync
+      image: argoproj/argocd
+      command: ["bash", "-c", "source ~/env.sh > /dev/null ; argocd app sync ${inputs.params.application-name} --revision ${inputs.params.revision} ${inputs.params.flags}"]
```
Comment (vdemeester) [inline argocd/argocd.yaml]: Same here
```
command: ["/bin/bash", "-c"]
args:
- source ~/env.sh > /dev/null ; argocd app sync ${inputs.params.application-name} --revision ${inputs.params.revision} ${inputs.params.flags}
```

File: argocd/argocd.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    - name: wait
+      image: argoproj/argocd
+      command: ["bash", "-c", "source ~/env.sh > /dev/null ; argocd app wait ${inputs.params.application-name} --health ${inputs.params.flags}"]
```
Comment (vdemeester) [inline argocd/argocd.yaml]: Same here
```
command: ["/bin/bash", "-c"]
args:
- source ~/env.sh > /dev/null ; argocd app wait ${inputs.params.application-name} --health ${inputs.params.flags}
```

**Comment by AdityaGupta1** [comment]
@vdemeester would something like the following work?

```YAML
apiVersion: v1
kind: ConfigMap
metadata:
  name: env-configmap
data:
  ARGOCD_SERVER: <server address>
---
apiVersion: v1
kind: Secret
metadata:
  name: env-secret
data:
  ARGOCD_USERNAME: <base64 encoded username>
  ARGOCD_PASSWORD: <base64 encoded password>
  ARGOCD_AUTH_TOKEN: <base64 encoded token or empty>
---
apiVersion: tekton.dev/v1alpha1
kind: Task
metadata:
  name: argocd-task-sync-and-wait
spec:
  inputs:
    params:
      - name: application-name
        description: name of the application to sync
      - name: revision
        description: the revision to sync to
        default: HEAD
      - name: flags
        default: --
  containerTemplate:
    envFrom:
      - configMapRef:
          name: env-configmap
      - secretRef:
          name: env-secret
  steps:
    - name: login
      image: argoproj/argocd
      command: ["/bin/bash", "-c"]
      args:
        - if [ -z $ARGOCD_AUTH_TOKEN ]; then
            yes | argocd login $ARGOCD_SERVER --username=ARGOCD_USERNAME --password=ARGOCD_PASSWORD;
          fi
    - name: sync
      image: argoproj/argocd
      command: ["/bin/bash", "-c"]
      args:
        - argocd app sync ${inputs.params.application-name} --revision ${inputs.params.revision} ${inputs.params.flags}
    - name: wait
      image: argoproj/argocd
      command: ["/bin/bash", "-c"]
      args:
        - argocd app wait ${inputs.params.application-name} --health ${inputs.params.flags}
```

If this is what works, I'll commit this along with an updated README with instructions on how to use the ConfigMap and Secret for authentication.

**Comment by dlorenc** [comment]
> If this is what works, I'll commit this along with an updated README with instructions on how to use the ConfigMap and Secret for authentication.

Yeah this makes sense!

**Comment by vdemeester** [comment]
```yaml
    envFrom:
      - configMapRef:
          name: env-configmap
      - secretRef:
          name: env-secret
```
Those could even be parametrize, aka having a parameter for the task and a default value, letting users use the secret name the want. But it will only work on `0.5.0` (as `containerTemplate` doesn't support template before https://github.com/tektoncd/pipeline/pull/1006 is merged).

**Comment by AdityaGupta1** [comment]
@dlorenc I've updated it to use `ConfigMap` and `Secret`, and the instructions have been updated to match

**Comment by AdityaGupta1** [comment]
It seems like there are some bugs with the latest version of Argo CD that prevent it from working, but my internship at Intuit ends today and I doubt I'll have time to work on it afterwards. Reverting to container version 1.0.2 solves the problem for username/password login, but there is still a bug with token login that is being looked into

**Comment by AdityaGupta1** [comment]
@vdemeester @dlorenc I just noticed there is a typo in the login command, in that it is missing a `$` before `ARGOCD_USERNAME` and `ARGOCD_PASSWORD`. Should I submit another pull request to fix that?

**Pull request merged**